src/coreclr/jit/gentree.cpp

```diff
@@ -24415,58 +24415,69 @@ GenTree* Compiler::gtNewSimdSumNode(var_types type, GenTree* op1, CorInfoType si
+    }
 
     if (simdSize == 32)
     {
-        // Minus 1 because for the last pass we split the vector to low / high and add them together.
-        haddCount -= 1;
+        assert(compIsaSupportedDebugOnly(InstructionSet_AVX2));
+        GenTree* op1Dup = fgMakeMultiUse(&op1);
+        op1             = gtNewSimdGetUpperNode(TYP_SIMD16, op1, simdBaseJitType, simdSize);
+        op1Dup          = gtNewSimdGetLowerNode(TYP_SIMD16, op1Dup, simdBaseJitType, simdSize);
+        simdSize        = simdSize / 2;
+        op1             = gtNewSimdBinOpNode(GT_ADD, TYP_SIMD16, op1, op1Dup, simdBaseJitType, simdSize);
+    }
 
-        if (varTypeIsFloating(simdBaseType))
+    assert(simdSize == 16);
+    if (varTypeIsFloating(simdBaseType))
+    {
+        GenTree* op1Dup = nullptr;
+        if (simdBaseType == TYP_FLOAT)
         {
-            assert(compIsaSupportedDebugOnly(InstructionSet_AVX));
-            intrinsic = NI_AVX_HorizontalAdd;
+            assert(compIsaSupportedDebugOnly(InstructionSet_SSE));
+            intrinsic = NI_SSE_Shuffle;
+            op1Dup    = fgMakeMultiUse(&op1);
+            op1 = gtNewSimdHWIntrinsicNode(TYP_SIMD16, op1, op1Dup, gtNewIconNode(177), intrinsic, simdBaseJitType, 16);
+            op1Dup = fgMakeMultiUse(&op1Dup);
+            op1    = gtNewSimdBinOpNode(GT_ADD, TYP_SIMD16, op1, op1Dup, simdBaseJitType, simdSize);
         }
         else
         {
-            assert(compIsaSupportedDebugOnly(InstructionSet_AVX2));
-            intrinsic = NI_AVX2_HorizontalAdd;
+            assert(compIsaSupportedDebugOnly(InstructionSet_SSE2));
+            intrinsic = NI_SSE2_Shuffle;
         }
-    }
-    else if (varTypeIsFloating(simdBaseType))
-    {
-        assert(compIsaSupportedDebugOnly(InstructionSet_SSE3));
-        intrinsic = NI_SSE3_HorizontalAdd;
-    }
-    else
-    {
-        assert(compIsaSupportedDebugOnly(InstructionSet_SSSE3));
-        intrinsic = NI_SSSE3_HorizontalAdd;
+        op1Dup = fgMakeMultiUse(&op1);
+        op1    = gtNewSimdHWIntrinsicNode(TYP_SIMD16, op1, op1Dup, gtNewIconNode(3), intrinsic, simdBaseJitType, 16);
+        op1Dup = fgMakeMultiUse(&op1Dup);
+        op1    = gtNewSimdBinOpNode(GT_ADD, TYP_SIMD16, op1, op1Dup, simdBaseJitType, simdSize);
+        return gtNewSimdHWIntrinsicNode(type, op1, NI_Vector128_ToScalar, simdBaseJitType, 16);
     }
 
-    for (int i = 0; i < haddCount; i++)
-    {
-        tmp = fgMakeMultiUse(&op1);
-        op1 = gtNewSimdHWIntrinsicNode(simdType, op1, tmp, intrinsic, simdBaseJitType, simdSize);
-    }
+    unsigned vectorLength = getSIMDVectorLength(simdSize, simdBaseType);
+    int      shiftCount   = genLog2(vectorLength);
+    int      typeSize     = genTypeSize(simdBaseType);
+    int      shiftVal     = (typeSize * vectorLength) / 2;
 
-    if (simdSize == 32)
+    GenTree* opShifted = nullptr;
+    while (shiftVal >= typeSize)
```

Comment:
This could likewise do with a small comment explaining how the logic here works. Doing a sum via shifting isn't necessarily "obvious" at a glance

Reply:
Added comment